### PR TITLE
Refactor UnivariatePolynomial to UnivariateIntPolynomial

### PR DESCRIPTION
**Pull request by chenchfort**
UnivariatePolynomial does not handle non-integer coefficients. Misleading.

### CODE REVIEW

File: symengine/polynomial.h [hunk 1]
```diff
@@ -85,20 +85,20 @@ class UnivariatePolynomial : public Basic{
...
+RCP<const UnivariateIntPolynomial> add_uni_poly(const UnivariateIntPolynomial &a, const UnivariateIntPolynomial &b);
+//! Negative of a UnivariateIntPolynomial
+RCP<const UnivariateIntPolynomial> neg_uni_poly(const UnivariateIntPolynomial &a);
+//! Subtracting two UnivariateIntPolynomial a and b
+RCP<const UnivariateIntPolynomial> sub_uni_poly(const UnivariateIntPolynomial &a, const UnivariateIntPolynomial &b);
+//! Multiplying two UnivariateIntPolynomial a and b
+RCP<const UnivariateIntPolynomial> mul_uni_poly(RCP<const UnivariateIntPolynomial> a, RCP<const UnivariateIntPolynomial> b);
 
-inline RCP<const UnivariatePolynomial> univariate_polynomial(RCP<const Symbol> i, unsigned int deg, map_uint_mpz&& dict)
+inline RCP<const UnivariateIntPolynomial> univariate_polynomial(RCP<const Symbol> i, unsigned int deg, map_uint_mpz&& dict)
```
Comment (isuruf) [inline symengine/polynomial.h]: This should be changed, right?

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -143,81 +143,81 @@ integer_class UnivariatePolynomial::eval_bit(const int &x) const {
...
     for (const auto &it : b.dict_)
         dict[it.first] += it.second;
 
-    RCP<const UnivariatePolynomial> c = univariate_polynomial(a.var_, (--(dict.end()))->first, std::move(dict));
+    RCP<const UnivariateIntPolynomial> c = univariate_polynomial(a.var_, (--(dict.end()))->first, std::move(dict));
     return c;
 }
 
-RCP<const UnivariatePolynomial> neg_uni_poly(const UnivariatePolynomial &a) {
+RCP<const UnivariateIntPolynomial> neg_uni_poly(const UnivariateIntPolynomial &a) {
```
Comment (isuruf) [inline symengine/polynomial.cpp]: I think these methods were changed to `neg_poly`, `add_poly` etc. right?

**Comment by isuruf** [comment]
Can you rebase on top of master?
`git rebase upstream/master`

**Comment by chenchfort** [comment]
Anything else?

**Comment by isuruf** [comment]
Aside from this, https://github.com/symengine/symengine/pull/851#discussion_r55115264, looks good to me.

File: symengine/polynomial.cpp (hunk 2 above)
Comment (chenchfort) [inline symengine/polynomial.cpp]: Yes, let me change that.

**Comment by isuruf** [comment]
This looks good. If you can rebase on top of master, we can get this merged.

**Comment by chenchfort** [comment]
Don't know exactly how rebase works. Execute git rebase upstream/master in this branch or master?

**Comment by isuruf** [comment]
In this branch

**Comment by chenchfort** [comment]
I just executed cmake -DCMAKE_BUILD_TYPE=Debug . and now it doesn't make.
It's complaining Users/Charles/symengine/symengine/symengine_rcp.cpp:4:10: fatal error: 'symengine/symengine_rcp.h' file not found

Any idea?

**Comment by isuruf** [comment]
Hmm, that's weird. Can you clean the repo (`git clean -dfx`) and try again?

**Comment by chenchfort** [comment]
I did
cmake .
make
Still same error.

**Comment by isuruf** [comment]
Did you by any chance delete that file? If not, I don't know why this error is there.

**Comment by chenchfort** [comment]
Nope.

**Comment by chenchfort** [comment]
Should I go ahead and delete the repo and clone it again?

**Comment by isuruf** [comment]
You can try,
`git clean -dfx`
`git reset --hard e859878`

**Comment by chenchfort** [comment]
Weird, still got that error...
In addition, when run cmake, it says,
No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.4)

**Comment by chenchfort** [comment]
I'm getting conflicts after doing rebase.

**Comment by isuruf** [comment]
After rebasing you should do a force push

**Comment by isuruf** [comment]
Looks good.
One final thing, you can also delete the first two commits which are not needed.
`git rebase -i HEAD~6`
Then delete 6094453 and c6039b6

**Comment by chenchfort** [comment]
There are some weird issues with some builds on Travis CI.

**Comment by isuruf** [comment]
Those are network issues. I restarted them

**Comment by chenchfort** [comment]
So instead of "pick", do I just change to "drop" for first two?

**Comment by isuruf** [comment]
+1 to merge after tests pass

**Comment by isuruf** [comment]
Thanks @chenchfort and team for the PR

**Comment by Sumith1896** [comment]
Sorry I've been off for a while. This looks good to me.
Thanks @chenchfort and team

**Pull request merged**